keps/sig-api-machinery/1904-efficient-watch-resumption/README.md

```diff
@@ -0,0 +1,485 @@
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+The kube-apiserver watch cache is initialized from etcd at the moment
+when it starts with empty transaction log. As a consequence, clients
+that want to resume a watch immediately after kube-apiserver reboots
+almost always have a resource version that is out of the history window.
+In this KEP we propose how to address this problem.
+
+## Motivation
+
+In order to support large Kubernetes clusters without huge number of etcd
+and kube-apiserver replicas, kube-apiserver contains an in-memory cache
+called `watch cache` from which all watch requests as well as non-quorum
+reads (in opt-in fashion) are served. Watch cache is being propagated
+directly from etcd using typical "list + watch" pattern and stores both
+the current state as well as som recent 'transaction log'. There is a
+separate watch cache for each resource type version.
+
+Digging into more details, the way how watch cache is propagated is:
+- on start, quorum list is send to etcd
+- the watch cache is set to be synced to resource version of that list
+- from that point, watch to etcd is established and watch cache is
+  updated by incoming watch event stream; note that only objects of
+  a given resource type are watched
+- if watch cannot be reestablished, all watchers connected to it are
+  disconnected, a new quorum list is send and watchcache is reset to
+  that point in time
+
+It's important to note that while only objects of a given resource type
+are watched (thus resource version to which watch cache is synced can
+only be updated to a value that some object of that type had), the quorum
+list, even though it is only requesting objects of that resource type,
+returns "globally current version for the whole etcd".
+
+While the inability to reestablish watch doesn't happen often in practice,
+the kube-apiserver restarts obviously does (e.g. on upgrades). And this
+is resulting in two main problems:
+
+1. In case of rolling upgrade of kube-apiserver, when no object of a given
+   resource type is changing (but objects of other types do), all watchers
+   will eventually be forced to relist, causing significant performance
+   and scalability issues for larger clusters. See the example below for
+   clarification.
+
+1. If no object of a given resource type is changing after kube-apiserver
+  startup, watch cache may stuck being synced to resource version not being
+	version of any object of a given type for extended period of time. See
+  https://github.com/kubernetes/kubernetes/issues/91073 for more details.
+
+To illustrate the problem better, imagine the following example:
+- there is a single kube-apiserver that we are going to restart
+- there are only two resource types: X and Y (for simplicity)
+- there is one object of type X: x with rv=100
+- there is one object of type Y: y with rv=101
```

Comment:
Thanks, this is much easier to follow now, I think you can also remove references to type Y?

Reply:
Actually it is needed, because I need some non-X-type object that is bumping be global resource version.